### PR TITLE
POS Cart redesign

### LINKED ISSUE #5109: POS Product Cart view - make it consistent with our design guidelines and dstruktify it!
Our POS Cart view seemed to be used quite a lot, however the design/styling of it looks quite outdated compared to our design system . This is a tracker issue that will try to log all the improvements we need to make, but those improvements will be posted here most likely  next week once we determine course of action.

Bear in mind that there will be category/group changes coming in that we should consider. #5078 
```[tasklist]
### Tasks
- [x] Come up with design @dstrukt
- [x] Implement proposed design
- [x] Ensure we only touch on cosmetic side of things, but also think if we'd like to improve the checkout flow for the next major release
```

### LINKED ISSUE #5130: Sticky header for POS Plugin
When a merchant has multiple items, it's currently difficult to navigate to the cart. The simplest solution would be to add sticky header.

https://github.com/btcpayserver/btcpayserver/assets/36959754/311ec8a1-30cf-487f-a179-cf4f83175bd4

### PR DESCRIPTION
**Pull request by dennisreimann**
Closes #5109. Closes #5130.

Quite a few things happened here:

- Updated the POS cart UI, as this was the purpose of the issue
- Removed the custom amount input for the cart version (as discussed in the dev and design meetings)
- Removed the modal confirmation step from the cart, as the initial overview is sufficient
- Refactored the POS Cart from jQuery to Vue.js
- Extracted a Vue mixin with common functionality which is now used by the keypad and cart
- Unified all POS views as much as possible, especially the item lists

I can imagine that we have another round of more code finetuning for the POS views and CSS, but that would be covered in a separate PR then.

## Item List

grafik

## Cart

grafik

### CODE REVIEW

**Comment by dstrukt** [comment]
HELL YEAH. Can't wait to review!

**Comment by Kevyke** [comment]
What is the site about

**Comment by dennisreimann** [comment]
Adapted it so that the cart is displayed as an offcanvas for mobile (see screenshots above) and sidebar for desktop (screenshots below).

dark

light

**Review by dstrukt — CHANGES_REQUESTED** [review]
## Still Reviewing, will remove when done.

## Overall, this is fantastic, so much better!!

Some of the easy adjustments I'll do / can help with, but documenting for both of us:

## List View

- [x] The title of the list and cart view should be a consistent font-size, looked and it seems to use the `<h1 class="store-name">` instead..
- [ ] Should we add search/categories to the List view?

Screen Shot 2023-07-20 at 1 43 50 PM

- [ ] Should we ellipses the content after X characters to prevent crazy differences in card height? It's fine here because the heights are pretty consistent, but could see it being an issue.
Screen Shot 2023-07-20 at 1 56 20 PM

## Cart View

### Desktop

- [x] We prevent the user from clicking past the inventory, but if you input greater than the inventory, shouldn't it just update back to the max? e.g. i'm able to input 8 with a 7 max, and i'm sure i'll hit an error, but it could be avoided with some JS, no?
Screen Shot 2023-07-20 at 1 50 03 PM

- [x] We might want to increase the width a little bit on the cart view, so we don't have wrapping issue? As much as I love 420 😎 maybe a bit more, 480? Open to other ideas here.
Screen Shot 2023-07-20 at 1 48 03 PM

- [x] Think we could just go with `All` vs. `All items` to save a little bit of horizontal space imo
Screen Shot 2023-07-20 at 1 47 43 PM

- [x] We should probably reduce the size of the cart icon from 40/40 to 32/32px, it's a bit large atm

- [x] We should also bump the badge font-size as well, it's currently too small, ideally 12pt or like `0.857...?em` or whatever the math is based on the default 14 pixel size

- [x] I'd love to be able to override the styling on these, the browser defaults are ugly
Screen Shot 2023-07-20 at 2 04 42 PM

- [x] Reduce the icon size to 2.25em from 2.5em
Screen Shot 2023-07-20 at 2 30 01 PM

- [ ] There is currently no way to close to the cart on desktop

- [x] I manage to break the number display when there is No quantity in the input, e.g: not sure if this requires action, but a note:
Screen Shot 2023-07-20 at 2 11 05 PM

- [ ] I changed the title halfway through editing, but it didn't update in the cart view, minor edge case, but an interesting one:
Screen Shot 2023-07-20 at 2 17 22 PM

- [x] Not 100% certain the table headers are necessary, so just thought I'd comment:
Screen Shot 2023-07-20 at 2 09 02 PM

- [ ] Similarly, I've removed the magnifying glass from some of the search inputs, but debating if it makes sense to add them back...

- [x] Update placeholder text from `Find product` to `Search...` or `Search for product...`. Thoughts?

- [x] We should bump right and left the padding on the inputs:
Screen Shot 2023-07-20 at 3 10 44 PM

- [x] Debating if we want to round the border radius to full on the input for this view specifically. It's a bit inconsistent from all inputs, but think it looks nice and differentiates search on specific views. Thoughts?
Screen Shot 2023-07-20 at 3 10 59 PM

- [x] Also still trying to decide if I like the `Empty Cart` at the bottom of the view, logically it does make some sense, but it could accidentally be pressed and be quite annoying. I had originally placed it to the right of the logo / title, which would work on desktop in the current form, but might be a bit messy on mobile.. hmmm

- [ ] From a functional standpoint, clicking both the card and the button adds the product to the cart, and this is a minor functional nit .. but should it just be limited to the button to add things to the cart?

### Mobile

- [x] The `X` icon is black, not white
- [x] Should ideally be horizontally aligned with the title of the store as well
Screen Shot 2023-07-20 at 2 06 23 PM

- [ ] Should we make the pills scrollable like we do with the tabbed navigation? Thoughts?
Screen Shot 2023-07-20 at 3 24 55 PM

- [ ] Should there be an "empty cart" state? Currently you just can't click into the cart if it's empty. Thoughts? Am able to get this view when clearing the cart and degrading to mobile though - this works:
Screen Shot 2023-07-20 at 3 34 00 PM

- [ ] Not 100% sure I like the slide-in animation for the cart on mobile, maybe just a dissolve in/out might be better. Thoughts? Side thought, but doing a 100% width view for the main navigation (as opposed to the the current 75-80%), and the animation would be treated the same, in either case.

## Additional Thoughts & Considerations:

- Do we want to include Store branding, logos, etc..? This might be a follow up PR, but it's something I could understand some users wanting.

**Comment by dennisreimann** [comment]
Thanks @dstrukt, I addressed most of the feedback, see the recent commits.

> Should we add search/categories to the List view?

This one does not have any JavaScript yet and isn't as dynamic as the cart, so that would be a separate overhaul I'd say.

>  Should we ellipses the content after X characters to prevent crazy differences in card height? It's fine here because the heights are pretty consistent, but could see it being an issue.

This behaves as before and so far this hasn't been brought up, so I'd say we are good for this PR. However, I think this is a valid point to address in a future update.

>  I'd love to be able to override the styling on these, the browser defaults are ugly

Let's do this in an additional round of updates as well, as I'm time constrained for the rest of this release.

> From a functional standpoint, clicking both the card and the button adds the product to the cart, and this is a minor functional nit .. but should it just be limited to the button to add things to the cart?

@pavlenex and I spoke about this and decided to keep the current behaviour to not change the flow for merchants. It's easier to click the whole item than just the button, which makes it better to handle on mobile/tablets.

File: BTCPayServer.Tests/SeleniumTests.cs [hunk 1]
```diff
@@ -2189,47 +2189,58 @@ public async Task CanUsePOSCart()
...
+            s.Driver.ElementDoesNotExist(By.Id("CartDiscount"));
+            s.Driver.FindElement(By.Id("Discount")).SendKeys("10");
+            Assert.Contains("10% = 0,30 €", s.Driver.FindElement(By.Id("CartDiscount")).Text);
+            Assert.Equal("2,70 €", s.Driver.FindElement(By.Id("CartTotal")).Text);
             
-            // Custom amount
-            s.Driver.FindElement(By.Id("CartCustomAmount")).SendKeys("1.5");
-            s.Driver.FindElement(By.Id("CartTotal")).Click();
-            Assert.Equal("3,50 €", s.Driver.FindElement(By.Id("CartTotal")).Text);
-            s.Driver.FindElement(By.Id("js-cart-confirm")).Click();
```
Comment (dennisreimann) [inline BTCPayServer.Tests/SeleniumTests.cs:2223]: Custom amount was emoved from the cart view as discussed.

File: BTCPayServer/Views/Shared/PosData.cshtml [hunk 2]
```diff
@@ -23,8 +23,7 @@
                     {
                         <th class="w-150px">@key</th>
                     }
-                    <td style="white-space:pre-wrap">
-                        @if (IsValidURL(str))
+                    <td style="white-space:pre-wrap">@* Explicitely remove whitespace at front here *@@if (IsValidURL(str))
```
Comment (dennisreimann) [inline BTCPayServer/Views/Shared/PosData.cshtml:26]: Otherwise the value column starts with an empty line and the data is shown from line 2 on.

File: BTCPayServer/Views/UIOnChainAutomatedPayoutProcessors/Configure.cshtml [hunk 3]
```diff
@@ -26,23 +26,23 @@
             <div class="form-group">
                 <label asp-for="IntervalMinutes" class="form-label" data-required>Interval</label>
                 <div class="input-group">
-                    <input asp-for="IntervalMinutes" class="form-control" inputmode="numeric" style="max-width:10ch;">
+                    <input asp-for="IntervalMinutes" class="form-control" inputmode="numeric" style="max-width:12ch;">
```
Comment (dennisreimann) [inline BTCPayServer/Views/UIOnChainAutomatedPayoutProcessors/Configure.cshtml:29]: These changes needed to be done as we bumped the horizontal padding of the input fields in the design system.

**Comment by dennisreimann** [comment]
Updated the cart display to incorporate @dstrukt's feedback/

dark

light

**Review by pavlenex — APPROVED** [review]
tACK, great work @dennisreimann and thanks @dstrukt for in-depth testing and feedback.

It looks great. My only comment is that on destkop view the icon cart still remains though it doesn't really do much. I like how this was handled for mobile and in generally very aesthetically pleasing. Good work team!

**Comment by dennisreimann** [comment]
I thought it's nice to have it with the badge, but we can obviously also remove it on desktop. Let us know what you think, @dstrukt

**Comment by dstrukt** [comment]
I see both sides, and agree with @dennisreimann that it looks nice with the amount badge, but think I expect it to open/close the cart as well, because that's how it functions on mobile.

Hmm, the lack of functionality on desktop doesn't bother me, and think we should just keep it for the time being / see if get any additional feedback that it's annoying or confusing.

Will do a proper round of review on the last set of changes, thanks for knocking these out so quickly @dennisreimann !!

**Comment by dstrukt** [comment]
> This one does not have any JavaScript yet and isn't as dynamic as the cart, so that would be a separate overhaul I'd say.

Sounds good, and agree, totally finally to leave out on this, wasn't sure how far the scope went. 

> This behaves as before and so far this hasn't been brought up, so I'd say we are good for this PR. However, I think this is a valid point to address in a future update.

Sounds good as well, an easy enough patch if the issue arises with more usage.
 
> Let's do this in an additional round of updates as well, as I'm time constrained for the rest of this release.

Sounds good, there's definitely some very minor design system stuff that I need to make a few issues for.
 
> @pavlenex and I spoke about this and decided to keep the current behaviour to not change the flow for merchants. It's easier to click the whole item than just the button, which makes it better to handle on mobile/tablets.

> We also talked about this and decided to leave out this functionality.

Sounds good wrt the former, will see if the lack of functionality on the icon, as discussed above, will be an issue.

**Comment by dstrukt** [comment]
## Final thoughts:

Overall, love the update, looks fantastic - there are a few design system nits that will get addressed in after this is merged, as well as improving the singular list view, so that will all be wrapped up nicely.

Last few nits, don't need them to be done, but if they are, great, and won't require a follow up review.

- [ ] Empty button should probably be green
Screen Shot 2023-07-21 at 12 02 34 PM

- [ ] Thought I made this change, but should be `All`, unless I misunderstood something?
Screen Shot 2023-07-21 at 12 02 29 PM

- [ ] Might be a separate follow up, but this button hover state should be a shade or two lighter than the default
Screen Shot 2023-07-21 at 12 02 22 PM

**Review by dstrukt — APPROVED** [review]
LEGGO!

**Comment by dstrukt** [comment]
If we remove the cart icon, don't we want to scale up the badge to 32px, and have a "0" state? Just saw the latest commits.

**Comment by pavlenex** [comment]
Alright, seems we can merge this one then @NicolasDorier 🚀

**Comment by dstrukt** [comment]
Ahh I see, it's just removed. LGTM!

File: BTCPayServer/Views/Shared/PointOfSale/Public/Cart.cshtml [hunk 4]
```diff
@@ -1,349 +1,244 @@
...
-            <a class="js-cart btn btn-sm bg-white text-black pull-right ms-5" href="#">
-                <i class="fa fa-times fa-lg"></i>
-            </a>
-            <a class="js-cart-destroy btn btn-danger pull-right" href="#" style="display: none;" id="CartClear">Empty cart <i class="fa fa-trash fa-fw fa-lg"></i></a>
+    <aside id="cart" ref="cart" tabindex="-1" aria-labelledby="cartLabel">
+        <div class="public-page-wrap" v-cloak>
+            <div class="container-xl">
+                <header class="sticky-top bg-tile offcanvas-header py-3 py-lg-4 d-flex align-items-baseline justify-content-center gap-3 px-5 pe-lg-0">
+                    <h1 class="mb-0" id="cartLabel">Cart</h1>
+                    <button id="CartClear" type="reset" v-on:click="clearCart" class="btn btn-text text-primary p-1">
```
Comment (dennisreimann) [inline BTCPayServer/Views/Shared/PointOfSale/Public/Cart.cshtml:125]:
```suggestion
                    <button id="CartClear" type="reset" v-on:click="clearCart" class="btn btn-text text-primary p-1" v-if="cartCount > 0">
```

**Comment by dennisreimann** [comment]
> Visual edge case to have the Empty button on ana empty cart though. Minor.

Fixed.

**Comment by dstrukt** [comment]
LGTM! Ready for merge.

File: BTCPayServer.Tests/SeleniumTests.cs [hunk 5]
```diff
@@ -2189,47 +2189,58 @@ public async Task CanUsePOSCart()
...
+            s.Driver.WaitForElement(By.Id("PosItems"));
+            s.Driver.ElementDoesNotExist(By.Id("CartBadge"));
+            Assert.Empty(s.Driver.FindElements(By.CssSelector("#CartItems tr")));
             
             // Select and clear
-            s.Driver.FindElement(By.CssSelector(".card.js-add-cart:nth-child(1)")).Click();
-            Assert.Single(s.Driver.FindElements(By.CssSelector("#js-cart-list tbody tr")));
+            s.Driver.FindElement(By.CssSelector(".posItem:nth-child(1) .btn-primary")).Click();
+            Assert.Single(s.Driver.FindElements(By.CssSelector("#CartItems tr")));
+            Assert.Equal("1", s.Driver.FindElement(By.Id("CartBadge")).Text);
```
Comment (dennisreimann) [inline BTCPayServer.Tests/SeleniumTests.cs:2207]:
```suggestion
```

File: BTCPayServer.Tests/SeleniumTests.cs [hunk 6]
```diff
@@ -2189,47 +2189,58 @@ public async Task CanUsePOSCart()
...
             s.Driver.FindElement(By.Id("CartClear")).Click();
-            Assert.Empty(s.Driver.FindElements(By.CssSelector("#js-cart-list tbody tr")));
             Thread.Sleep(250);
+            Assert.Empty(s.Driver.FindElements(By.CssSelector("#CartItems tr")));
+            s.Driver.ElementDoesNotExist(By.Id("CartBadge"));
```
Comment (dennisreimann) [inline BTCPayServer.Tests/SeleniumTests.cs:2211]:
```suggestion
```

File: BTCPayServer.Tests/SeleniumTests.cs [hunk 7]
```diff
@@ -2189,47 +2189,58 @@ public async Task CanUsePOSCart()
...
             s.Driver.FindElement(By.Id("ViewApp")).Click();
             var windows = s.Driver.WindowHandles;
             Assert.Equal(2, windows.Count);
             s.Driver.SwitchTo().Window(windows[1]);
-            s.Driver.WaitForElement(By.Id("js-cart-list"));
-            Assert.Empty(s.Driver.FindElements(By.CssSelector("#js-cart-list tbody tr")));
-            Assert.Equal("0,00 €", s.Driver.FindElement(By.Id("CartTotal")).Text);
-            Assert.False(s.Driver.FindElement(By.Id("CartClear")).Displayed);
+            s.Driver.WaitForElement(By.Id("PosItems"));
+            s.Driver.ElementDoesNotExist(By.Id("CartBadge"));
```
Comment (dennisreimann) [inline BTCPayServer.Tests/SeleniumTests.cs:2201]:
```suggestion
```

File: BTCPayServer.Tests/SeleniumTests.cs [hunk 8]
```diff
@@ -2189,47 +2189,58 @@ public async Task CanUsePOSCart()
...
             Thread.Sleep(250);
+            Assert.Empty(s.Driver.FindElements(By.CssSelector("#CartItems tr")));
+            s.Driver.ElementDoesNotExist(By.Id("CartBadge"));
             
             // Select items
-            s.Driver.FindElement(By.CssSelector(".card.js-add-cart:nth-child(2)")).Click();
+            s.Driver.FindElement(By.CssSelector(".posItem:nth-child(2) .btn-primary")).Click();
             Thread.Sleep(250);
-            s.Driver.FindElement(By.CssSelector(".card.js-add-cart:nth-child(1)")).Click();
+            Assert.Equal("1", s.Driver.FindElement(By.Id("CartBadge")).Text);
```
Comment (dennisreimann) [inline BTCPayServer.Tests/SeleniumTests.cs:2216]:
```suggestion
```

File: BTCPayServer.Tests/SeleniumTests.cs [hunk 9]
```diff
@@ -2189,47 +2189,58 @@ public async Task CanUsePOSCart()
...
+            s.Driver.FindElement(By.CssSelector(".posItem:nth-child(1) .btn-primary")).Click();
             Thread.Sleep(250);
-            Assert.Equal(2, s.Driver.FindElements(By.CssSelector("#js-cart-list tbody tr")).Count);
-            Assert.Equal("2,00 €", s.Driver.FindElement(By.Id("CartTotal")).Text);
+            Assert.Equal("2", s.Driver.FindElement(By.Id("CartBadge")).Text);
```
Comment (dennisreimann) [inline BTCPayServer.Tests/SeleniumTests.cs:2219]:
```suggestion
```

File: BTCPayServer.Tests/SeleniumTests.cs [hunk 10]
```diff
@@ -2189,47 +2189,58 @@ public async Task CanUsePOSCart()
...
+            s.Driver.FindElement(By.CssSelector(".posItem:nth-child(2) .btn-primary")).Click();
+            Thread.Sleep(250);
+            Assert.Equal("3", s.Driver.FindElement(By.Id("CartBadge")).Text);
```
Comment (dennisreimann) [inline BTCPayServer.Tests/SeleniumTests.cs:2222]:
```suggestion
```
